### PR TITLE
Allow partial result in minimal_model, and translate OS-algebra to cdga

### LINKED ISSUE #35781: Allow minimal model of cdga to return partial results
### Is there an existing issue for this?

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.

### Problem Description

Minimal model of cdga's might be infinitely generated, so we haver a limit on the iteration steps tried. If this limit is reached, we return an error.

However, in some situations (for instance, studying Orlik-Solomon algebras of hyperplane arrangements), it can be interested to get the result computed up to a given point (even if it is not really an i-minimal model).

### Proposed Solution

To implement a `partial_result` flag in the  mimimal model method. This flag should be set to `False` by default, but when enabled, it would return the partially computed result instead of an error.

### Alternatives Considered

I see no other sensible alternative.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by miguelmarco**
### :books: Description
Fixes #35781 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/algebras/commutative_dga.py [hunk 1]
```diff
@@ -2699,17 +2730,28 @@ def extendy(phi, degree):
             B = A.cdg_algebra(A.differential({}))
             # Solve case that fails with one generator return B,gens
             phi = B.hom(gens)
-            phi = extendy(phi, degnzero + 1)
+            phiext = extendy(phi, degnzero + 1)
+            if isinstance(phiext, tuple):
+                if partial_result:
+                    return phiext[0]
+                else:
+                    raise ValueError("could not cover all relations in max iterations in degree {}".format(degnzero + 1))
```
Comment (tscrim) [inline src/sage/algebras/commutative_dga.py]: A more personal idiosyncrasy but
```suggestion
                if not partial_result:
                    raise ValueError("could not cover all relations in max iterations in degree {}".format(degnzero + 1))
                return phiext[0]
```
I find to be slightly easier to understand (or just removing the `else:`) as it more clearly indicates that it does terminate.

File: src/sage/algebras/orlik_solomon.py [hunk 2]
```diff
@@ -446,6 +446,33 @@ def degree_on_basis(self, m):
         """
         return len(m)
 
+    def as_cdga(self):
+        """
+        Return the commutative differential graded algebra corresponding to self, with
+        trivial differential.
```
Comment (tscrim) [inline src/sage/algebras/orlik_solomon.py]:
```suggestion
        r"""
        Return the commutative differential graded algebra corresponding to ``self``
        with the trivial differential.
```

File: src/sage/algebras/orlik_solomon.py [hunk 3]
```diff
@@ -446,6 +446,33 @@ def degree_on_basis(self, m):
...
+               e2 --> 0
+        """
+        from sage.algebras.commutative_dga import GradedCommutativeAlgebra
+        gens = list(self.algebra_generators())
+        names = ['e{}'.format(i) for i in range(len(gens))]
+        A = GradedCommutativeAlgebra(self.base_ring(), names)
+        bas2 = [(i,j) for j in range(len(gens)) for i in range(j) if gens[i]*gens[j] in self.basis()]
+        non_basis = [(i,j) for j in range(len(gens)) for i in range(j) if not gens[i]*gens[j] in self.basis()]
+        rels = {(i,j) : {p : (gens[i]*gens[j]).coefficient(frozenset(p)) for p in bas2} for (i,j) in non_basis}
+        I = A.ideal([A.gen(l)*A.gen(m) - sum(k*A.gen(i)*A.gen(j) for ((i,j),k) in rels[(l,m)].items()) for (l,m) in non_basis])
```
Comment (tscrim) [inline src/sage/algebras/orlik_solomon.py]:
```suggestion
        B = self.basis()
        bas2 = []
        non_basis = []
        for j, gj in enumerate(gens):
            for i in range(j):
                if gens[i] * gj in B:
                    bas2.append((i, j))
                else:
                    non_basis.append((i, j))
        AG = A.gens()
        AGbas = {(i, j): AG[i] * AG[j] for (i, j) in bas2}
        gij = {(i, j): gens[i] * gens[j] for (i, j) in bas2}

        def rhs(l, m):
            P = gens[l] * gens[m]
            return A.sum(P.coefficient(gij[p]) * AGbas[p] for p in bas2)

        I = A.ideal([AG[l]*AG[m] - rhs(l, m) for (l,m) in non_basis])
```
I tried as much as I could to remove repeated products as I would expect those to be slow.

**Review by tscrim — CHANGES_REQUESTED** [review]
Why is the method to return the OS algebra as differential algebra and not just a CGA? It seems like that would be the more natural thing to have as the user could then easily enough add the differential.

**Comment by miguelmarco** [comment]
> Why is the method to return the OS algebra as differential algebra and not just a CGA? It seems like that would be the more natural thing to have as the user could then easily enough add the differential.

The OS algebra is the cohomology ring of the complement of the arrangement, so i guess it makes sense to consider it as a differential algebra.

But as you said, since it is so easy to add the differential structure, it is almost equivalent to return it just as gca.

File: src/sage/algebras/orlik_solomon.py (hunk 2 above)
Comment (miguelmarco) [reply src/sage/algebras/orlik_solomon.py]: Sounds good.

File: src/sage/algebras/orlik_solomon.py (hunk 3 above)
Comment (miguelmarco) [reply src/sage/algebras/orlik_solomon.py]: I don't think there will be a siginifcant speed difference, but anyways, your approach is probably more legible.

**Comment by miguelmarco** [comment]
> 



> 



> Perhaps we can have both methods, but with the `gdca` simply calling the `gca()` and adding the differential structure?

Yes, that makes sense.

File: src/sage/algebras/commutative_dga.py [hunk 4]
```diff
@@ -2576,6 +2581,32 @@ def minimal_model(self, i=3, max_iterations=3):
...
+               y1_0 --> x1_0*x1_1 - x1_0*x1_2 + x1_1*x1_2
+               y1_1 --> x1_0*y1_0 - x1_2*y1_0
+               y1_2 --> x1_1*y1_0 - x1_2*y1_0
+              To:   Commutative Differential Graded Algebra with generators ('a', 'b', 'c') in degrees (1, 1, 1) with relations [a*b - a*c + b*c] over Rational Field with differential:
+               a --> 0
+               b --> 0
+               c --> 0
+              Defn: (x1_0, x1_1, x1_2, y1_0, y1_1, y1_2) --> (a, b, c, 0, 0, 0)
+
+
```
Comment (tscrim) [inline src/sage/algebras/commutative_dga.py]:
```suggestion

```
The linter could get slightly mad for too many blank lines...

File: src/sage/algebras/orlik_solomon.py [hunk 5]
```diff
@@ -446,6 +446,59 @@ def degree_on_basis(self, m):
         """
         return len(m)
 
+    def as_gca(self):
+        """
+        Return the graded commutative algebra corresponding to self.
```
Comment (tscrim) [inline src/sage/algebras/orlik_solomon.py]:
```suggestion
        r"""
        Return the graded commutative algebra corresponding to ``self``.
```

File: src/sage/algebras/orlik_solomon.py [hunk 6]
```diff
@@ -446,6 +446,59 @@ def degree_on_basis(self, m):
...
+
+        EXAMPLES::
+
+            sage: H = hyperplane_arrangements.braid(3)
+            sage: O = H.orlik_solomon_algebra(QQ)
+            sage: O.as_gca()
+            Graded Commutative Algebra with generators ('e0', 'e1', 'e2') in degrees (1, 1, 1) with relations [e0*e1 - e0*e2 + e1*e2] over Rational Field
```
Comment (tscrim) [inline src/sage/algebras/orlik_solomon.py]:
```suggestion
            Graded Commutative Algebra with generators ('e0', 'e1', 'e2') in degrees (1, 1, 1)
             with relations [e0*e1 - e0*e2 + e1*e2] over Rational Field
```

**Review by tscrim — CHANGES_REQUESTED** [review]
Thanks. Just a few other essentially trivial things.

File: src/sage/algebras/orlik_solomon.py (hunk 2 above)
Comment (tscrim) [reply src/sage/algebras/orlik_solomon.py]: Once this comment is addressed, then it will be a positive review.

**Comment by miguelmarco** [comment]
So, is it ready for positive review?

File: src/sage/algebras/commutative_dga.py [hunk 7]
```diff
@@ -2437,7 +2437,12 @@ def minimal_model(self, i=3, max_iterations=3):
 
         - ``max_iterations`` -- integer (default: `3`); the number of
           iterations of the method at each degree. If the algorithm does not
-          finish in this many iterations at each degree, an error is raised.
+          finish in this many iterations at each degree, an error is raised,
+          or the partial result computed up to that point is returned, deppending
+          on the `partial_result` flag.
+
+        - ``partial_result``  -- boolean (default: `False`); wether to return
+          the partial result if the `max_iterations` limit is reached.
```
Comment (tscrim) [inline src/sage/algebras/commutative_dga.py]:
```suggestion
          on the ``partial_result`` flag.

        - ``partial_result``  -- boolean (default: ``False``); wether to return
          the partial result if the ``max_iterations`` limit is reached
```

File: src/sage/algebras/commutative_dga.py [hunk 8]
```diff
@@ -2576,6 +2581,31 @@ def minimal_model(self, i=3, max_iterations=3):
...
+            sage: I = A.ideal([a*b-a*c+b*c])
+            sage: B = A.quotient(I)
+            sage: S = B.cdg_algebra({})
+            sage: S.minimal_model()
+            Traceback (most recent call last):
+            ...
+            ValueError: could not cover all relations in max iterations in degree 2
+            sage: S.minimal_model(partial_result=True)
+            Commutative Differential Graded Algebra morphism:
+              From: Commutative Differential Graded Algebra with generators ('x1_0', 'x1_1', 'x1_2', 'y1_0', 'y1_1', 'y1_2') in degrees (1, 1, 1, 1, 1, 1) over Rational Field with differential:
```
Comment (tscrim) [inline src/sage/algebras/commutative_dga.py]:
```suggestion
              From: Commutative Differential Graded Algebra with generators
               ('x1_0', 'x1_1', 'x1_2', 'y1_0', 'y1_1', 'y1_2') in degrees (1, 1, 1, 1, 1, 1)
                over Rational Field with differential:
```

File: src/sage/algebras/commutative_dga.py [hunk 9]
```diff
@@ -2576,6 +2581,31 @@ def minimal_model(self, i=3, max_iterations=3):
...
+               x1_0 --> 0
+               x1_1 --> 0
+               x1_2 --> 0
+               y1_0 --> x1_0*x1_1 - x1_0*x1_2 + x1_1*x1_2
+               y1_1 --> x1_0*y1_0 - x1_2*y1_0
+               y1_2 --> x1_1*y1_0 - x1_2*y1_0
+              To:   Commutative Differential Graded Algebra with generators ('a', 'b', 'c') in degrees (1, 1, 1) with relations [a*b - a*c + b*c] over Rational Field with differential:
```
Comment (tscrim) [inline src/sage/algebras/commutative_dga.py]:
```suggestion
              To:   Commutative Differential Graded Algebra with generators ('a', 'b', 'c')
               in degrees (1, 1, 1) with relations [a*b - a*c + b*c] over Rational Field with differential:
```

File: src/sage/algebras/orlik_solomon.py [hunk 10]
```diff
@@ -446,6 +446,71 @@ def degree_on_basis(self, m):
...
+             with relations [e0*e1 - e0*e2 + e1*e2] over Rational Field
+
+        ::
+
+            sage: N = matroids.named_matroids.Fano()
+            sage: O = N.orlik_solomon_algebra(QQ)
+            sage: O.as_gca()
+            Graded Commutative Algebra with generators ('e0', 'e1', 'e2', 'e3', 'e4', 'e5', 'e6') in degrees (1, 1, 1, 1, 1, 1, 1) with relations [e1*e2 - e1*e3 + e2*e3, e0*e2 - e0*e4 + e2*e4, e0*e1 - e0*e5 + e1*e5, e3*e4 - e3*e5 + e4*e5, e0*e3 - e0*e6 + e3*e6, e1*e4 - e1*e6 + e4*e6, e2*e5 - e2*e6 + e5*e6] over Rational Field
+
+        """
```
Comment (tscrim) [inline src/sage/algebras/orlik_solomon.py]:
```suggestion
            Graded Commutative Algebra with generators ('e0', 'e1', 'e2', 'e3', 'e4', 'e5', 'e6')
             in degrees (1, 1, 1, 1, 1, 1, 1) with relations
             [e1*e2 - e1*e3 + e2*e3, e0*e2 - e0*e4 + e2*e4, e0*e1 - e0*e5 + e1*e5,
              e3*e4 - e3*e5 + e4*e5, e0*e3 - e0*e6 + e3*e6, e1*e4 - e1*e6 + e4*e6,
              e2*e5 - e2*e6 + e5*e6] over Rational Field
        """
```
```

File: src/sage/algebras/orlik_solomon.py [hunk 11]
```diff
@@ -446,6 +446,71 @@ def degree_on_basis(self, m):
...
+        r"""
+        Return the commutative differential graded algebra corresponding to ``self``
+        with the trivial differential.
+
+        EXAMPLES::
+
+            sage: H = hyperplane_arrangements.braid(3)
+            sage: O = H.orlik_solomon_algebra(QQ)
+            sage: O.as_cdga()
+            Commutative Differential Graded Algebra with generators ('e0', 'e1', 'e2') in degrees (1, 1, 1) with relations [e0*e1 - e0*e2 + e1*e2] over Rational Field with differential:
```
Comment (tscrim) [inline src/sage/algebras/orlik_solomon.py]:
```suggestion
            Commutative Differential Graded Algebra with generators ('e0', 'e1', 'e2') in degrees
             (1, 1, 1) with relations [e0*e1 - e0*e2 + e1*e2] over Rational Field with differential:
```

**Review by tscrim — APPROVED** [review]
Sorry, just a few more trivial things I noticed while looking at the compiled doc. One formatting thing, and then breaking up the long output lines (might even be a bit better to be slightly shorter per line too; I didn't check closely).

I've approved the PR, so once you push the changes, you can switch the status to positive review.

**Comment by tscrim** [comment]
Good catch. Is there some way we can easily/quickly test that the result matches, say, (partially) computing the (graded) degrees?

**Comment by miguelmarco** [comment]
We can check if the Betti numbers match the coefficients of the poincare polynomial. For instance:

```
sage: H = hyperplane_arrangements.Catalan(4,QQ)
sage: O = H.cone().orlik_solomon_algebra(QQ)
sage: B = O.as_cdga()
sage: H.cone().poincare_polynomial()
210*x^4 + 317*x^3 + 125*x^2 + 19*x + 1
sage: [len(B.basis(i)) for i in range(5)]
[1, 19, 125, 317, 210]
```

```
sage: H = hyperplane_arrangements.Ish(5,QQ)
sage: O = H.cone().orlik_solomon_algebra(QQ)
sage: B = O.as_gca()
sage: H.cone().poincare_polynomial()
625*x^5 + 1125*x^4 + 650*x^3 + 170*x^2 + 21*x + 1
sage: [len(B.basis(i)) for i in range(6)]
[1, 21, 170, 650, 1125, 625]
```

**Comment by tscrim** [comment]
If that’s a quick (in the doctest time sense), then can we add that as an additional test? I promise, absolutely last thing. `^^;;`

**Comment by miguelmarco** [comment]
> If that’s a quick (in the doctest time sense), then can we add that as an additional test? I promise, absolutely last thing. `^^;;`

Done

**Comment by tscrim** [comment]
Thank you.

**Pull request merged**